<a href="https://colab.research.google.com/github/Ammara-Hussain/flyrank-internship-assignments/blob/main/assignments/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/lara-taan/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

My Chosen Path:
Lane 2: Refresh / Content Opportunity Scoring.

Why I chose this:
I am choosing this path because it perfectly matches the core goal of the project: fixing old or dying web pages. The starter code and practice dataset provided in the repository are already set up for this lane, making it the smartest place to start. By focusing on this, I can build a system that looks at real search traffic data (like how old a page is or how many views it gets) and automatically ranks those pages so the team knows exactly which ones to update first.

I chose the Refresh / Content Opportunity Scoring lane.

The goal is to help content teams decide which pages should be reviewed first for possible improvement. Instead of manually reviewing thousands of pages, this project will create a ranked list of content items using observable search and engagement signals. The output will support human decision-making by highlighting pages with potential opportunities for refresh, optimization, or monitoring.

This lane is valuable because content teams have limited time and need a data-driven way to prioritize where to focus their effort.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

The Search Question: Which web pages are highly visible but are actively dying (losing traffic), making them the highest priority for our team to look at first?

The Unit of Analysis: A single web page belonging to a specific client.

The Decision & Action: An editor needs to decide how to spend their limited working hours. The action they take is opening a page from the top of the list and updating its content or metadata to stop it from losing readers.

The Cost of a Wrong Call:

If the code makes a mistake and ranks a healthy page at the top (False Positive): The team wastes precious hours editing an article that didn't need to be touched.

If the code misses a dying page (False Negative): That page keeps losing readers every single day, and our competitors take over our traffic.

Why Data/ML helps: We have hundreds of thousands of web pages to track. It is humanly impossible for a person to check them all manually. Machine learning can automatically weigh multiple signals (views, age, drops in rank) to rank the list instantly and objectively.

The decision this project improves is: which content pages should be reviewed first for potential refresh opportunities.

The person acting on the output is a content strategist or SEO specialist. They can use the ranked list to decide which pages need investigation, such as updating content, improving engagement, or monitoring performance.

A wrong recommendation can waste editing time on pages that do not need improvement, while missing pages that have real opportunities. Therefore, the system should prioritize accurate ranking and provide explanations for why a page was selected.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [ ]:
import pandas as pd

# Load the starter dataset directly from the public GitHub URL
url = "https://raw.githubusercontent.com/flyrank-bih/flyrank-ml-internship-starter/main/data/raw/content_refresh_anonymized.csv"
df = pd.read_csv(url)

# Print columns to verify exactly what they are named
print("Columns found in dataset:", list(df.columns))
print("-" * 50)

# 1. Total unique pages we are dealing with
# Look for 'content_id' or 'content_hash_id' automatically
id_col = 'content_id' if 'content_id' in df.columns else 'content_hash_id'
if id_col in df.columns:
    total_pages = df[id_col].nunique()
    print(f"Total Unique Pages in Slice: {total_pages}")
else:
    print(f"Total rows in dataset: {len(df)}")

# 2. How many are actively trending downward (losing traffic)
# Look for 'trend_direction' or 'trend' automatically
trend_col = [c for c in df.columns if 'trend' in c.lower()]
if trend_col:
    print(f"\nTrend Direction Counts ({trend_col[0]}):")
    print(df[trend_col[0]].value_counts())

# 3. Pages with high visibility but low click rates
imp_col = [c for c in df.columns if 'impress' in c.lower()]
ctr_col = [c for c in df.columns if 'ctr' in c.lower()]

if imp_col and ctr_col:
    visible_low_ctr = df[(df[imp_col[0]] >= 500) & (df[ctr_col[0]] < 0.05)].shape[0]
    print(f"\nVisible pages (>=500 impressions) with low CTR (<5%): {visible_low_ctr}")
else:
    print("\nCould not calculate CTR metrics due to missing impressions/ctr columns.")

Columns found in dataset: ['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']
--------------------------------------------------
Total Unique Pages in Slice: 30000

Trend Direction Counts (trend_direction):
trend_direction
down      16262
stable     5962
up         4388
new       

The dataset contains 30,000 anonymized content items across 32 clients.

The starter dataset includes observable search and engagement signals such as impressions, CTR, average position, sessions, and trend measurements.

A first observation is that content performance is not uniform: pages have different trend directions, meaning some pages are improving while others are declining. This supports the need for a ranking system that helps prioritize review efforts.

In [ ]:
declining_rate = (df["trend_direction"] == "down").mean()

print("Percentage of declining pages:", round(declining_rate * 100, 2), "%")

Percentage of declining pages: 54.21 %


Around 54.21% of pages show a downward trend in the dataset, indicating that identifying which pages require attention is a meaningful prioritization problem.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

What my work WILL be able to say:

Observed Metrics: This system will provide clear, historical data tracking which specific pages have experienced drops in impressions or click-through rates.

Directional Trends: The model will identify the general direction a page's traffic is moving (e.g., separating pages that are stably growing from those that are actively decaying).

Decision-Support: The final output acts purely as a prioritized recommendation engine to help human content editors figure out where to spend their limited manual review hours first.

What my work NEVER will be able to say:

Causal Proof: This model cannot prove why a page is dropping, nor can it guarantee that updating the page will instantly cause its traffic to bounce back.

"Predicting Google": This work will never claim to reverse-engineer or predict Google’s secret search engine algorithm updates. It only measures the observable aftermath of search behavior on our own pages.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.